### ISSUE TITLE
LDAPS requests to ActiveDirectory server hang until timeout

### ISSUE THREAD

**Issue by Wolfgang29** [curl/curl#9580]
LDAPS requests to ActiveDirectory server hang until timeout
This seems only affect ActiveDirectory, LDAPS requests to OpenLDAP server works fine.

operating system: Windows 10, Windows Server 2016
Different Active Directory Server versions, all using AD mode "Win2012 R2"

libcurl version 7.85.0 (same befaviour with 7.84.0)

libcurl compiled on Windows 10 with VisualStudio 17.
Using: OpenSSL 1.1.1q, OpenLDAP 2.4.48

curl/7.85.0 libcurl/7.85.0 OpenSSL/1.1.1q libssh2/1.8.0 OpenLDAP/2.4.48

With the previous used libcurl version 7.69.1 all LDAPS requests works fine. (build with the same OpenSSL and OpenLDAP version)

Example of a used LDAPS request:

ldaps://ipaddress/dc=value1,dc=value2,dc=value2??sub?(&(objectClass=*)(sAMAccountName=*))

It seems, that the requestet data where received and then the end of the data transfer is not recognized. It hangs until timeout:

My result of the request (timeout 60 seconds): 
Operation timed out after 60006 milliseconds with 632348 bytes received

The same request without ssl are successful in a few seconds.

**Comment by bagder** [comment]
Unfortunately, LDAP(S) is not being verified in the curl test suite at all while also being little used, so regressions happen. I don't even know an LDAPS server to try your URL against in order to try to reproduce.

**Comment by Wolfgang29** [comment]
ich you can give me an ip address from which you will connect, than I can give you access to one of our test AD server.


Mit freundlichen Grüßen / Best regards,​
Wolfgang Kruck
Software Development
USU Solutions GmbH
Stadionstraße 4-6
70771 Leinfelden-Echterdingen
Telefon: +49 711 94707 45
Mobil: -
***@***.***
www.usu.com
USU Solutions GmbH, Sitz: Leinfelden-Echterdingen, Registergericht und Handelsregisternummer: Amtsgericht Stuttgart HRB 224464,
​Geschäftsführer: Claus Mohoric · Holger Sampel

**Comment by henning-schild** [comment]
Maybe try another ssl implementation for that client. i.e. take a Linux and try different precompiled curls with docker.

Alone finding an affected pre-compiled binary will allow you to start talking to Microsoft as well and give them a way to repro without a need to compile "your binary".

And you might get more insights on which combination broke things and which component is really to blame. Might not be curl at all.

Increase the log level of that server and dig into the logs. Maybe take a decrypted pcap. That can later be shown to others that might want to help, and you can compare the network packages from the "good old" with the "bad new" and maybe get a clue what might confuse that server.

**Comment by henning-schild** [comment]
And given you have a working and a not-working version you can simply "git bisect" and might be able to say which commit introduced your problem. Maybe that commit will give the clue on what might be going on.

**Comment by Wolfgang29** [comment]
I will debug it, as soon as I have time

**Comment by Wolfgang29** [comment]
New information about the problem (testet with the cURL version 7.87.0:

It seems, that LDAP and LDAPS only works, if the server closes the connection after the request.

If I make a LDAP request to the server, then the server closes the connection after the request and cURL gets the response.
If I make the same request as LDAPS request, then the server sends the response (all expectet data where received), but the server let the connection open. cURL waits in this situation until the configured timeout.

It seems that cURL depends on the connection close of the server. I think thats wrong. The end of the LDAP search request is the LDAP searchResDone message. This should be the end of the request and not the end of the connection.

**Comment by Wolfgang29** [comment]
I've debugged the problem and I can tell you more details of what happens.

Here a summary of a Wireshark trace which contains the problem:

TLSv1.2	571	Client Hello
TLSv1.2	247	Server Hello, Change Cipher Spec, Finished
TLSv1.2	161	Change Cipher Spec, Finished
LDAP	219	bindRequest(1) "CN=*************,CN=*****,dc=************,dc=******,dc=*****" simple 
LDAP	155	bindResponse(1) success 
LDAP	251	searchRequest(2) "dc=************,dc=******,dc=*****" wholeSubtree 
LDAP	355	searchResEntry(2) "CN=*************,CN=*****,DC=************,DC=******,DC=*****"  | searchResRef(2)  | searchResRef(2)  | searchResRef(2)  | searchResDone(2) success  [1 result]

In the first three lines of the trace the ssl connection is established.
Then a LDAP bindRequest with bindResponse success follows.
Then the LDAP searchRequest is sent.
The response of the searchRequest is a packet with 5 LDAP packets: 
The search result searchResEntry, followed by three referrals searchResRef, followed by the successful end of the search request searchResDone.

I had a breakpoint in the file openldap.c at the beginning of the function oldap_recv() and checkt the return value of the call 
rc = ldap_result(li->ld, lr->msgid, LDAP_MSG_ONE, &tv, &msg);

ldap_result() rc -> 0x61   LDAP_RES_BIND           bindResponse
ldap_result() rc -> 0x64   LDAP_RES_SEARCH_ENTRY   searchResEntry

after the first paket (searchResEntry) of the 5 packets of the last response, oldap_recv() and ldap_result() are never called.
The last 4 packets of the result remain in the LDAP queue.
There are continous calls of Curl_readwrite(), but with no effect. The call dont check the already received data.

Only if the timout occur or if I manually close the connection with Sysinternals Tcpview, than the last packets in the 
LDAP queue are reached

ldap_result() rc -> 0x73   LDAP_RES_SEARCH_REFERENCE
ldap_result() rc -> 0x73   LDAP_RES_SEARCH_REFERENCE
ldap_result() rc -> 0x73   LDAP_RES_SEARCH_REFERENCE
ldap_result() rc -> 0x65   LDAP_RES_SEARCH_RESULT

I think the real problem is, that if a ldap packet contains more than one ldap packets, than only the first is processed.

I dont know the cURL code good enough to fix the problem. I hope, the information helps you to fix the problem.

**Comment by henning-schild** [comment]
I might be able to run such a call against an AD LDAP setup i have access to, if i could repro i could maybe help.

Can you please provide the exact cmdline you are using, for the bits you might have to mask please point out what example values might be. And please include all context such as kerberos `kinit` or whatever might be needed to auth. In my setup i would need to use kerberos.

And maybe make sure to format code blocks with markdown with https://www.markdownguide.org/extended-syntax/#fenced-code-blocks and edit the comments you already did send.

**Comment by Wolfgang29** [comment]
I found the problem that caused the hanging ldaps connection in the file transfer.c in the function Curl_readwrite().

The function seems only to check if there are data pending on the socket and ignores already received data. So I forced the call of the function readwrite_data() with the following line:

``` c
  select_res |= CURL_CSELECT_IN;
```

I know, that is probably not the best solution, but it solves the problem with hanging ldaps connections.
Please consider the use of this change or make a similar change that checks also the received buffers and not 
only the pending data on the socket.

Here is the complete changed function. I've marked my change with ! Change START ! and ! Change END !

``` c
CURLcode Curl_readwrite(struct connectdata *conn,
                        struct Curl_easy *data,
                        bool *done,
                        bool *comeback)
{
  struct SingleRequest *k = &data->req;
  CURLcode result;
  int didwhat = 0;

  curl_socket_t fd_read;
  curl_socket_t fd_write;
  int select_res = conn->cselect_bits;

  conn->cselect_bits = 0;

  /* only use the proper socket if the *_HOLD bit is not set simultaneously as
     then we are in rate limiting state in that transfer direction */

  if((k->keepon & KEEP_RECVBITS) == KEEP_RECV)
    fd_read = conn->sockfd;
  else
    fd_read = CURL_SOCKET_BAD;

  if((k->keepon & KEEP_SENDBITS) == KEEP_SEND)
    fd_write = conn->writesockfd;
  else
    fd_write = CURL_SOCKET_BAD;

#if defined(USE_HTTP2) || defined(USE_HTTP3)
  if(data->state.drain) {
    select_res |= CURL_CSELECT_IN;
    DEBUGF(infof(data, "Curl_readwrite: forcibly told to drain data"));
  }
#endif

  if(!select_res) /* Call for select()/poll() only, if read/write/error
                     status is not known. */
    select_res = Curl_socket_check(fd_read, CURL_SOCKET_BAD, fd_write, 0);

  if(select_res == CURL_CSELECT_ERR) {
    failf(data, "select/poll returned error");
    result = CURLE_SEND_ERROR;
    goto out;
  }

#ifdef USE_HYPER
  if(conn->datastream) {
    result = conn->datastream(data, conn, &didwhat, done, select_res);
    if(result || *done)
      goto out;
  }
  else {
#endif

  /* ! Change START ! @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
     The functions above does not check if there are data in some buffers, in case of 
     LDAPS requests to an active directory server and the use of OpenLDAP and OpenSSL
     the data in the buffers are only checked within the readwrite_data function below.
     so the call of the readwrite_data function is forced. If the call of the 
     readwrite_data function is not forced, then the connection can hang until timeout.   */

  select_res |= CURL_CSELECT_IN;

  /* ! Change END ! @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ */

  /* We go ahead and do a read if we have a readable socket or if
     the stream was rewound (in which case we have data in a
     buffer) */

  if((k->keepon & KEEP_RECV) && (select_res & CURL_CSELECT_IN)) {
    result = readwrite_data(data, conn, k, &didwhat, done, comeback);
    if(result || *done)
      goto out;
  }

  /* If we still have writing to do, we check if we have a writable socket. */
  if((k->keepon & KEEP_SEND) && (select_res & CURL_CSELECT_OUT)) {
    /* write */

    result = readwrite_upload(data, conn, &didwhat);
    if(result)
      goto out;
  }
#ifdef USE_HYPER
  }
#endif

  k->now = Curl_now();
  if(!didwhat) {
    /* no read no write, this is a timeout? */
    if(k->exp100 == EXP100_AWAITING_CONTINUE) {
      /* This should allow some time for the header to arrive, but only a
         very short time as otherwise it'll be too much wasted time too
         often. */

      /* Quoting RFC2616, section "8.2.3 Use of the 100 (Continue) Status":

         Therefore, when a client sends this header field to an origin server
         (possibly via a proxy) from which it has never seen a 100 (Continue)
         status, the client SHOULD NOT wait for an indefinite period before
         sending the request body.

      */

      timediff_t ms = Curl_timediff(k->now, k->start100);
      if(ms >= data->set.expect_100_timeout) {
        /* we've waited long enough, continue anyway */
        k->exp100 = EXP100_SEND_DATA;
        k->keepon |= KEEP_SEND;
        Curl_expire_done(data, EXPIRE_100_TIMEOUT);
        infof(data, "Done waiting for 100-continue");
      }
    }

#ifdef ENABLE_QUIC
    if(conn->transport == TRNSPRT_QUIC) {
      result = Curl_quic_idle(data);
      if(result)
        goto out;
    }
#endif
  }

  if(Curl_pgrsUpdate(data))
    result = CURLE_ABORTED_BY_CALLBACK;
  else
    result = Curl_speedcheck(data, k->now);
  if(result)
    goto out;

  if(k->keepon) {
    if(0 > Curl_timeleft(data, &k->now, FALSE)) {
      if(k->size != -1) {
        failf(data, "Operation timed out after %" CURL_FORMAT_TIMEDIFF_T
              " milliseconds with %" CURL_FORMAT_CURL_OFF_T " out of %"
              CURL_FORMAT_CURL_OFF_T " bytes received",
              Curl_timediff(k->now, data->progress.t_startsingle),
              k->bytecount, k->size);
      }
      else {
        failf(data, "Operation timed out after %" CURL_FORMAT_TIMEDIFF_T
              " milliseconds with %" CURL_FORMAT_CURL_OFF_T " bytes received",
              Curl_timediff(k->now, data->progress.t_startsingle),
              k->bytecount);
      }
      result = CURLE_OPERATION_TIMEDOUT;
      goto out;
    }
  }
  else {
    /*
     * The transfer has been performed. Just make some general checks before
     * returning.
     */

    if(!(data->req.no_body) && (k->size != -1) &&
       (k->bytecount != k->size) &&
#ifdef CURL_DO_LINEEND_CONV
       /* Most FTP servers don't adjust their file SIZE response for CRLFs,
          so we'll check to see if the discrepancy can be explained
          by the number of CRLFs we've changed to LFs.
       */
       (k->bytecount != (k->size + data->state.crlf_conversions)) &&
#endif /* CURL_DO_LINEEND_CONV */
       !k->newurl) {
      failf(data, "transfer closed with %" CURL_FORMAT_CURL_OFF_T
            " bytes remaining to read", k->size - k->bytecount);
      result = CURLE_PARTIAL_FILE;
      goto out;
    }
    if(!(data->req.no_body) && k->chunk &&
       (conn->chunk.state != CHUNK_STOP)) {
      /*
       * In chunked mode, return an error if the connection is closed prior to
       * the empty (terminating) chunk is read.
       *
       * The condition above used to check for
       * conn->proto.http->chunk.datasize != 0 which is true after reading
       * *any* chunk, not just the empty chunk.
       *
       */
      failf(data, "transfer closed with outstanding read data remaining");
      result = CURLE_PARTIAL_FILE;
      goto out;
    }
    if(Curl_pgrsUpdate(data)) {
      result = CURLE_ABORTED_BY_CALLBACK;
      goto out;
    }
  }

  /* Now update the "done" boolean we return */
  *done = (0 == (k->keepon&(KEEP_RECV|KEEP_SEND|
                            KEEP_RECV_PAUSE|KEEP_SEND_PAUSE))) ? TRUE : FALSE;
  result = CURLE_OK;
out:
  DEBUGF(infof(data, "Curl_readwrite(handle=%p) -> %d", data, result));
  return result;
}
```

**Comment by henning-schild** [comment]
Cool. I guess it would make some sense to make that a commit to push to a fork and open a pull-request in this repo. That might get the whole story some attention and which patch gets merged in the end remains to be seen. Dumping diffs and code into comments of an issue will not have that same impact.

**Comment by bagder** [comment]
I would suspect that maybe `data_pending()` is the actual function that needs attention to drain the data properly...

**Comment by ozancsl** [comment]
Is there any proceed about this issue? I am having the same problem as well. I am planning to use integrate active-directory to my application with curl but this issue blocks me.

**Comment by bagder** [comment]
Our `KNOWN_BUGS` document contains 67 issues right now. They are listed there because they are bugs we know exist in current curl. That list includes this issue. I have not seen anyone even try to work on this yet. Feel free to do so!

**Comment by claires20080808** [comment]
@bagder  i met this issue in 8.6.0 version again. Could you please let me know whether this issue is fixed in latest version? In 8.9.1 ? Or is there any plan to fix in future version?

Thank you very much.

**Comment by bagder** [comment]
I have no idea. I don't have any way to reproduce the problem and then consequently I also have no way to verify if any change fixed it. But I don't think anyone deliberately have fixed this.

> Or is there any plan to fix in future version?

Ideally all problems get fixed. But this is one of those issues that probably needs someone who can reproduce the problem to "step up" and debug it proper - or provide a recipe that lets others to repro the case. Until that happens, it probably remains unfixed.

**Comment by andreiaugustin** [comment]
Sorry to revive this, please let me know if you'd prefer a new issue.

We're experiencing problems with LDAPS as well and think it may be related to this bug.

Things seem to improve when using curl 8.11.0 through the CLI, as in it does not hang anymore, but when used through the C API the problem is still there (which seems weird, I'd have expected the C API to render the same results as the CLI).

Here's my builds:

```
[jenkins@6200e9102c92 bin]$ /curl8110-original --version
curl 8.11.0 (x86_64-pc-linux-gnu) libcurl/8.11.0 (mbedTLS/3.5.2) OpenSSL/3.0.11 zlib/1.3.1 zstd/1.4.4 libidn2/2.2.0 libssh2/1.11.0 nghttp2/1.59.0 OpenLDAP/2.4.46
Release-Date: 2024-11-06
Protocols: file ftp ftps http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s scp sftp smtp smtps ws wss
Features: alt-svc AsynchDNS HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz MultiSSL NTLM SSL threadsafe TLS-SRP UnixSockets zstd
```

```
[jenkins@6200e9102c92 bin]$ /curl860 --version
curl 8.6.0 (x86_64-pc-linux-gnu) libcurl/8.6.0 (mbedTLS/3.5.2) OpenSSL/3.0.11 zlib/1.3.1 zstd/1.4.4 libidn2/2.2.0 libssh2/1.11.0 nghttp2/1.59.0 OpenLDAP/2.4.46
Release-Date: 2024-01-31
Protocols: file ftp ftps http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s scp sftp smtp smtps
Features: alt-svc AsynchDNS HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz MultiSSL NTLM SSL threadsafe TLS-SRP UnixSockets zstd
```

Running on RHEL8, with `export CURL_SSL_BACKEND=openssl`.

I've spent some time to set up an LDAPS server for testing purposes (was not a nice experience at all): 

`--url "ldaps://ldaps.omnis.net:636/DC=omnis,DC=net"`

`--user "omnis\curl:7Sg3HfP9RTG93v"`

Please note I am not a Windows Server expert and therefore don't know whether querying the above DC stuff is the right way to go about it, but I can see data coming back to the client which I think suffices for this bug!

Let's start with the CLI, using curl 8.6.0:

```
[jenkins@6200e9102c92 bin]$ /curl860 -vvvv --user "omnis\curl:7Sg3HfP9RTG93v" --url "ldaps://ldaps.omnis.net:636/DC=omnis,DC=net" -m 5
* Host ldaps.omnis.net:636 was resolved.
* IPv6: (none)
* IPv4: 13.41.119.29
*   Trying 13.41.119.29:636...
* Connected to ldaps.omnis.net (13.41.119.29) port 636
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/pki/tls/certs/ca-bundle.crt
*  CApath: none
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / X25519 / RSASSA-PSS
* Server certificate:
*  subject: C=GB; L=Saxmundham; O=Omnis Software Ltd; CN=*.omnis.net
*  start date: Jul 30 00:00:00 2024 GMT
*  expire date: Aug 14 23:59:59 2025 GMT
*  subjectAltName: host "ldaps.omnis.net" matched cert's "*.omnis.net"
*  issuer: C=US; O=DigiCert Inc; CN=DigiCert Global G2 TLS RSA SHA256 2020 CA1
*  SSL certificate verify ok.
*   Certificate level 0: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 2: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* LDAP local: ldaps://ldaps.omnis.net:636/DC=omnis,DC=net
DN: DC=omnis,DC=net
        objectClass: top
        objectClass: domain
        objectClass: domainDNS

        distinguishedName: DC=omnis,DC=net

        instanceType: 5

        whenCreated: 20241204150719.0Z

        whenChanged: 20241204162043.0Z

        subRefs: CN=Configuration,DC=omnis,DC=net

        uSNCreated: 4099

        dSASignature:: AQAAACgAAAAAAAAAAAAAAAAAAAAAAAAArUW/kM9Uu0OZdOFtcye2Nw==

        uSNChanged: 16407

        name: omnis

        objectGUID:: QPICfLbnKkqGRp6MVHthHw==

        creationTime: 133777985761123267

        forceLogoff: -9223372036854775808

        lockoutDuration: -6000000000

        lockOutObservationWindow: -6000000000

        lockoutThreshold: 0

        maxPwdAge: -36288000000000

        minPwdAge: 0

        minPwdLength: 0

        modifiedCountAtLastProm: 0

        nextRid: 1000

        pwdProperties: 1

        pwdHistoryLength: 0

        objectSid:: AQQAAAAAAAUVAAAAFuj+8WJiI8aWqb7M

        serverState: 1

        uASCompat: 0

        modifiedCount: 1

        auditingPolicy:: AAE=

        nTMixedDomain: 0

        rIDManagerReference: CN=RID Manager$,CN=System,DC=omnis,DC=net

        fSMORoleOwner: CN=NTDS Settings,CN=EC2AMAZ-TQL9KHQ,CN=Servers,CN=Default-First-Site-Name,CN=Sites,CN=Configuration,DC=omnis,DC=net

        systemFlags: -1946157056

        wellKnownObjects: B:32:6227F0AF1FC2410D8E3BB10615BB5B0F:CN=NTDS Quotas,DC=omnis,DC=net
        wellKnownObjects: B:32:F4BE92A4C777485E878E9421D53087DB:CN=Microsoft,CN=Program Data,DC=omnis,DC=net
        wellKnownObjects: B:32:09460C08AE1E4A4EA0F64AEE7DAA1E5A:CN=Program Data,DC=omnis,DC=net
        wellKnownObjects: B:32:22B70C67D56E4EFB91E9300FCA3DC1AA:CN=ForeignSecurityPrincipals,DC=omnis,DC=net
        wellKnownObjects: B:32:18E2EA80684F11D2B9AA00C04F79F805:CN=Deleted Objects,DC=omnis,DC=net
        wellKnownObjects: B:32:2FBAC1870ADE11D297C400C04FD8D5CD:CN=Infrastructure,DC=omnis,DC=net
        wellKnownObjects: B:32:AB8153B7768811D1ADED00C04FD8D5CD:CN=LostAndFound,DC=omnis,DC=net
        wellKnownObjects: B:32:AB1D30F3768811D1ADED00C04FD8D5CD:CN=System,DC=omnis,DC=net
        wellKnownObjects: B:32:A361B2FFFFD211D1AA4B00C04FD7D83A:OU=Domain Controllers,DC=omnis,DC=net
        wellKnownObjects: B:32:AA312825768811D1ADED00C04FD8D5CD:CN=Computers,DC=omnis,DC=net
        wellKnownObjects: B:32:A9D1CA15768811D1ADED00C04FD8D5CD:CN=Users,DC=omnis,DC=net

        objectCategory: CN=Domain-DNS,CN=Schema,CN=Configuration,DC=omnis,DC=net

        isCriticalSystemObject: TRUE

        gPLink: [LDAP://CN={31B2F340-016D-11D2-945F-00C04FB984F9},CN=Policies,CN=System,DC=omnis,DC=net;0]

        dSCorePropagationData: 16010101000000.0Z

        otherWellKnownObjects: B:32:683A24E2E8164BD3AF86AC3C2CF3F981:CN=Keys,DC=omnis,DC=net
        otherWellKnownObjects: B:32:1EB93889E40C45DF9F0C64D23BBB6237:CN=Managed Service Accounts,DC=omnis,DC=net

        masteredBy: CN=NTDS Settings,CN=EC2AMAZ-TQL9KHQ,CN=Servers,CN=Default-First-Site-Name,CN=Sites,CN=Configuration,DC=omnis,DC=net

        ms-DS-MachineAccountQuota: 10

        msDS-Behavior-Version: 10

        msDS-PerUserTrustQuota: 1

        msDS-AllUsersTrustQuota: 1000

        msDS-PerUserTrustTombstonesQuota: 10

        msDs-masteredBy: CN=NTDS Settings,CN=EC2AMAZ-TQL9KHQ,CN=Servers,CN=Default-First-Site-Name,CN=Sites,CN=Configuration,DC=omnis,DC=net

        msDS-IsDomainFor: CN=NTDS Settings,CN=EC2AMAZ-TQL9KHQ,CN=Servers,CN=Default-First-Site-Name,CN=Sites,CN=Configuration,DC=omnis,DC=net

        msDS-NcType: 0

        msDS-ExpirePasswordsOnSmartCardOnlyAccounts: TRUE

        dc: omnis


* Operation timed out after 5001 milliseconds with 3179 bytes received
* Closing connection
* Send failure: Bad file descriptor
* Recv failure: Bad file descriptor
curl: (28) Operation timed out after 5001 milliseconds with 3179 bytes received
```

And then the CLI using 8.11.0:

```
[jenkins@6200e9102c92 bin]$ /curl8110-original -vvvv --user "omnis\curl:7Sg3HfP9RTG93v" --url "ldaps://ldaps.omnis.net:636/DC=omnis,DC=net" -m 5 > log.txt 2>&1

16:38:47.484566 [0-x] == Info: [READ] client_reset, clear readers
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     016:38:47.537179 [0-0] == Info: Host ldaps.omnis.net:636 was resolved.
16:38:47.537478 [0-0] == Info: IPv6: (none)
16:38:47.537608 [0-0] == Info: IPv4: 13.41.119.29
16:38:47.537721 [0-0] == Info: [SETUP] added
16:38:47.537811 [0-0] == Info: [HAPPY-EYEBALLS] created ipv4 (timeout 4947ms)
16:38:47.537960 [0-0] == Info: [HAPPY-EYEBALLS] ipv4 starting (timeout=4947ms)
16:38:47.538120 [0-0] == Info:   Trying 13.41.119.29:636...
16:38:47.538252 [0-0] == Info: [TCP] cf_socket_open() -> 0, fd=5
16:38:47.538423 [0-0] == Info: [TCP] local address 172.17.0.2 port 53676...
16:38:47.538584 [0-0] == Info: [HAPPY-EYEBALLS] ipv4 connect -> 0, connected=0
16:38:47.538738 [0-0] == Info: [TCP] adjust_pollset, !connected, POLLOUT fd=5
16:38:47.538886 [0-0] == Info: [HAPPY-EYEBALLS] adjust_pollset -> 1 socks
16:38:47.549108 [0-0] == Info: [TCP] not connected yet
16:38:47.549232 [0-0] == Info: [HAPPY-EYEBALLS] ipv4 connect -> 0, connected=0
16:38:47.549383 [0-0] == Info: [TCP] adjust_pollset, !connected, POLLOUT fd=5
16:38:47.549535 [0-0] == Info: [HAPPY-EYEBALLS] adjust_pollset -> 1 socks
16:38:47.588758 [0-0] == Info: [TCP] not connected yet
16:38:47.588950 [0-0] == Info: [HAPPY-EYEBALLS] ipv4 connect -> 0, connected=0
16:38:47.589183 [0-0] == Info: [TCP] adjust_pollset, !connected, POLLOUT fd=5
16:38:47.589411 [0-0] == Info: [HAPPY-EYEBALLS] adjust_pollset -> 1 socks
16:38:47.625355 [0-0] == Info: [TCP] connected
16:38:47.625510 [0-0] == Info: [HAPPY-EYEBALLS] ipv4 connect -> 0, connected=1
16:38:47.625662 [0-0] == Info: [HAPPY-EYEBALLS] Connected to ldaps.omnis.net (13.41.119.29) port 636
16:38:47.625854 [0-0] == Info: [SSL] cf_connect()
16:38:47.627064 [0-0] == Info: [SSL] No cached session ID for ldaps://ldaps.omnis.net:636
16:38:47.627366 [0-0] => Send SSL data, 5 bytes (0x5)
0000: .....
16:38:47.627548 [0-0] == Info: TLSv1.3 (OUT), TLS handshake, Client hello (1):
16:38:47.627700 [0-0] => Send SSL data, 512 bytes (0x200)
0000: ......|.-A.-...........#}i!.N..4EK..]. ......>"..=Mf...3...J..e.
0040: K.......>.......,.0.........+./...$.(.k.#.'.g.....9.....3.....=.
0080: <.5./.....u.........ldaps.omnis.net.............................
00c0: ..............1.....*.(.........................................
0100: +............-.....3.&.$... :.'kn..>.....`.ft..1.....&.F..+{....
0140: ................................................................
0180: ................................................................
01c0: ................................................................
16:38:47.628940 [0-0] == Info: [TCP] send(len=517) -> 517, err=0
16:38:47.629099 [0-0] == Info: [SSL] ossl_bio_cf_out_write(len=517) -> 517, err=0
16:38:47.629266 [0-0] == Info: [TCP] recv(len=5) -> -1, err=81
16:38:47.629387 [0-0] == Info: [SSL] ossl_bio_cf_in_read(len=5) -> -1, err=81
16:38:47.629536 [0-0] == Info: [SSL] populate_x509_store, path=/etc/pki/tls/certs/ca-bundle.crt, blob=0
16:38:47.640699 [0-0] == Info:  CAfile: /etc/pki/tls/certs/ca-bundle.crt
16:38:47.640855 [0-0] == Info:  CApath: none
16:38:47.640945 [0-0] == Info: [SSL] SSL_connect() -> err=-1, detail=2
16:38:47.641078 [0-0] == Info: [SSL] SSL_connect() -> want recv
16:38:47.641198 [0-0] == Info: [SSL] cf_connect() -> 0, done=0
16:38:47.641325 [0-0] == Info: [SSL] adjust_pollset, POLLIN fd=5
16:38:47.641453 [0-0] == Info: [TCP] adjust_pollset, !active, POLLIN fd=5
16:38:47.717829 [0-0] == Info: [SSL] cf_connect()
16:38:47.717979 [0-0] == Info: [TCP] recv(len=5) -> 5, err=0
16:38:47.718100 [0-0] == Info: [SSL] ossl_bio_cf_in_read(len=5) -> 5, err=0
16:38:47.718245 [0-0] <= Recv SSL data, 5 bytes (0x5)
0000: ....z
16:38:47.718373 [0-0] == Info: [TCP] recv(len=122) -> 122, err=0
16:38:47.718497 [0-0] == Info: [SSL] ossl_bio_cf_in_read(len=122) -> 122, err=0
16:38:47.718667 [0-0] == Info: TLSv1.3 (IN), TLS handshake, Server hello (2):
16:38:47.718826 [0-0] <= Recv SSL data, 122 bytes (0x7a)
0000: ...v....s...A.e\...,a.Xv.w.N..5......o ......>"..=Mf...3...J..e.

[MANY LINES LATER]

16:40:35.234731 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.234847 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.235018 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.235237 [0-0] <= Recv data, 1 bytes (0x1)
0000: .
16:40:35.235361 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.235479 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.235666 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.235882 [0-0] <= Recv data, 2 bytes (0x2)
0000: dc
16:40:35.236024 [0-0] == Info: [WRITE] cw_out, wrote 2 body bytes -> 2
16:40:35.236205 [0-0] == Info: [WRITE] download_write body(type=1, blen=2) -> 0
16:40:35.236354 [0-0] == Info: [WRITE] client_write(type=1, len=2) -> 0
16:40:35.236487 [0-0] <= Recv data, 1 bytes (0x1)
0000: :
16:40:35.236606 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.236721 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.236893 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.237076 [0-0] <= Recv data, 1 bytes (0x1)
0000:  
16:40:35.237202 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.237364 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.237550 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.237725 [0-0] <= Recv data, 5 bytes (0x5)
0000: omnis
16:40:35.237922 [0-0] == Info: [WRITE] cw_out, wrote 5 body bytes -> 5
16:40:35.238109 [0-0] == Info: [WRITE] download_write body(type=1, blen=5) -> 0
16:40:35.238337 [0-0] == Info: [WRITE] client_write(type=1, len=5) -> 0
16:40:35.238424 [0-0] <= Recv data, 1 bytes (0x1)
0000: .
        dc: omnis
16:40:35.238573 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.238749 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.238930 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.239104 [0-0] <= Recv data, 1 bytes (0x1)
0000: .

16:40:35.239236 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.239399 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.239554 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.239706 [0-0] <= Recv data, 1 bytes (0x1)
0000: .

16:40:35.239809 [0-0] == Info: [WRITE] cw_out, wrote 1 body bytes -> 1
16:40:35.239990 [0-0] == Info: [WRITE] download_write body(type=1, blen=1) -> 0
16:40:35.240137 [0-0] == Info: [WRITE] client_write(type=1, len=1) -> 0
16:40:35.240298 [0-0] == Info: [WRITE] cw-out is notpaused
16:40:35.240466 [0-0] == Info: [SSL] cf_recv(len=8) -> 8, 0
16:40:35.240587 [0-0] == Info: [SSL] cf_recv(len=14) -> 14, 0
16:40:35.240723 [0-0] == Info: abort upload
16:40:35.240815 [0-0] <= Recv data, 0 bytes (0x0)
16:40:35.240945 [0-0] == Info: [WRITE] download_write body(type=81, blen=0) -> 0
16:40:35.241133 [0-0] == Info: [WRITE] client_write(type=81, len=0) -> 0
16:40:35.241263 [0-0] == Info: [WRITE] xfer_write_resp(len=0, eos=1) -> 0
16:40:35.241428 [0-0] == Info: [WRITE] cw-out is notpaused
16:40:35.241620 [0-0] == Info: [WRITE] cw-out done
16:40:35.241737 [0-0] == Info: [READ] client_reset, clear readers
16:40:35.241897 [0-0] == Info: Connection #0 to host ldaps.omnis.net left intact
```

Evidently things look better in the 8.11.0 CLI, although I get the same results as 8.6.0 when using the C API (namely bad file descriptor for the socket, as probably the timeout closed them?).

Over the next few days I will try an older version as our customers confirmed this happened after we upgraded to curl 8.6.0, and hopefully we can find a needle in the haystack that causes the hang through diffing.

Otherwise, I will start investigating the latest source and see if I can get lucky. In the meantime, if anybody else with more experience is willing to have a go, feel free to use the test server I've set up with the details aforementioned!

**Comment by dfandrich** [comment]
> Evidently things look better in the 8.11.0 CLI, although I get the same results
> as 8.6.0 when using the C API

Are you absolutely sure your C program is using the correct (newer) libcurl in this case?

**Comment by andreiaugustin** [comment]
Thanks @dfandrich ! It indeed was linking to the wrong libcurl.

I can still see a couple bad file descriptors in my logs which I don't see when using the CLI, but it does not hang until timeout anymore:

```
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 18:14:56.480 abort upload0
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 18:14:56.480 shutting down connection #00
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 0000  17 03 03 00 1f                                  .....0
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 18:14:56.480 Recv failure: Bad file descriptor0
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 0000  15 03 03 00 1a                                  .....0
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 18:14:56.480 Send failure: Bad file descriptor0
@@@ 20241204_181456: Info - oBL_LDAPonOW3.$runQuery -- ldap log: 18:14:56.480 Recv failure: Bad file descriptor0
```

I can confirm, using the latest cURL release at time of writing (8.11.0) to communicate over LDAPS with a Windows Server 2025 running ActiveDirectory Domain Services does not hang anymore for both CLI and C API!

I will leave the server running for a few more days in case somebody else would like to try it to confirm this is not a bug anymore and perhaps remove it from docs/KNOWN_BUGS (12.4) ?
